util/ffmpeg.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2022 Sumner Evans
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package util
+
+import (
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	log "maunium.net/go/maulogger/v2"
+)
+
+var ffmpegDefaultParams = []string{"-hide_banner", "-loglevel", "warning"}
+
+// MimeExtensionSanityOverrides includes extensions for various common mimetypes.
+//
+// This is necessary because sometimes the OS mimetype database and Go interact in weird ways,
+// which causes very obscure extensions to be first in the array for common mimetypes
+// (e.g. image/jpeg -> .jpe, text/plain -> ,v).
+var MimeExtensionSanityOverrides = map[string]string{
+	"image/png":  ".png",
+	"image/webp": ".webp",
+	"image/jpeg": ".jpg",
+	"image/tiff": ".tiff",
+	"image/heif": ".heic",
+	"image/heic": ".heic",
+
+	"audio/mpeg": ".mp3",
+	"audio/ogg":  ".ogg",
+	"audio/webm": ".webm",
+	"video/mp4":  ".mp4",
+	"video/mpeg": ".mpeg",
+	"video/webm": ".webm",
+
+	"text/plain": ".txt",
+	"text/html":  ".html",
+
+	"application/xml": ".xml",
+}
+
+func ExtensionFromMimetype(mimetype string) string {
+	ext, ok := MimeExtensionSanityOverrides[strings.Split(mimetype, ";")[0]]
+	if !ok {
+		exts, _ := mime.ExtensionsByType(mimetype)
+		if len(exts) > 0 {
+			ext = exts[0]
+		}
+	}
+	return ext
+}
+
+// Convert a media file on the disk using ffmpeg.
+//
+// Args:
+// * inputFile: The full path to the file.
+// * outputExtension: The extension that the output file should be.
+// * inputArgs: Arguments to tell ffmpeg how to parse the input file.
+// * outputArgs: Arguments to tell ffmpeg how to convert the file to reach the wanted output.
+// * removeInput: Whether the input file should be removed after converting.
+//
+// Returns: the path to the converted file.
+func ConvertPath(inputFile string, outputExtension string, inputArgs []string, outputArgs []string, removeInput bool) (string, error) {
+	outputFilename := strings.TrimSuffix(inputFile, filepath.Ext(inputFile)) + "." + outputExtension
+
+	args := []string{}
+	args = append(args, ffmpegDefaultParams...)
+	args = append(args, inputArgs...)
+	args = append(args, "-i", inputFile)
+	args = append(args, outputArgs...)
+	args = append(args, outputFilename)
+
+	log.Info(args)
+
+	cmd := exec.Command("ffmpeg", args...)
+	vcLog := log.Sub("ffmpeg").Writer(log.LevelWarn)
+	cmd.Stdout = vcLog
+	cmd.Stderr = vcLog
+	err := cmd.Run()
+	if err != nil {
+		return "", fmt.Errorf("ffmpeg error: %+v. stdout: %s. stderr: %s", err, cmd.Stdout, cmd.Stderr)
```

Comment:
I don't think this works, it just returns an error message containing pointers to the log writer. `cmd.StdoutPipe()` would work for getting the stream data as errors, but it might be easier to just return the code and look at the logs if something goes wrong

```suggestion
		return "", fmt.Errorf("ffmpeg error: %w", err)
```